### ISSUE TITLE
[flexAttention] compiled flexAttention with sliding window runs in 32bit float precision but crashes on 16bit float & mixed precision

### ISSUE THREAD

**Issue by cathalobrien** [pytorch/pytorch#133431]
### 🐛 Describe the bug

Hello, I'm using torch.compile on flexAttn with sliding window. When I run my model in 16-mixed or 16bit float precision the attention function segfaults during the initial backward pass. if I run in 32 bit float precision it runs, but clearly this isn't ideal. Ideally I could run in 16-mixed precision globally. I've tried auto casting the attention function and its overarching forward layer to float16 but I get the same segfault during the backward pass. 

below ive provided a repro, it runs in float32 precision but segfaults in float16

```python
import torch

from torch.nn.attention.flex_attention import flex_attention, create_block_mask

import functools

if __name__ == "__main__":

    B, H, SEQ_LEN, HEAD_DIM = 1, 16, 40320, 16 
    WINDOW_SIZE = 512

    def make_tensor():
        return torch.ones(B, H, SEQ_LEN, HEAD_DIM, device="cuda", dtype=torch.float16, requires_grad=True)

    q, k, v = make_tensor(), make_tensor(), make_tensor()
    gradOut = torch.ones(B, H, SEQ_LEN, HEAD_DIM, device="cuda", dtype=torch.float16)

    def sliding_window(b, h, q_idx, kv_idx):
            return torch.abs(q_idx - kv_idx) <= WINDOW_SIZE

    block_mask = create_block_mask(
        sliding_window, B=None, H=None, Q_LEN=SEQ_LEN, KV_LEN=SEQ_LEN, _compile=True
    )
    
    attention = functools.partial(flex_attention, block_mask=block_mask) #cache the block mask so its not remade

   
    attention = torch.compile(attention, fullgraph=False)

    out = attention(q, k, v, block_mask=block_mask)
    print(f"Shape of output tensor: {list(out.shape)}")
    out.backward(gradOut, retain_graph=True)
    print(f"Shape of output tensor after bw: {list(out.shape)}")
```

### Versions

Collecting environment information...
PyTorch version: 2.5.0.dev20240812+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Red Hat Enterprise Linux release 8.8 (Ootpa) (x86_64)
GCC version: (ECMWF) 12.2.0
Clang version: 15.0.7 (Red Hat 15.0.7-1.module+el8.8.0+17939+b58878af)
CMake version: Could not collect
Libc version: glibc-2.28

Python version: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-4.18.0-477.43.1.el8_8.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 12.3.103
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-SXM4-40GB
Nvidia driver version: 550.54.14
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.8.9.7
/usr/lib64/libcudnn_adv_infer.so.8.9.7
/usr/lib64/libcudnn_adv_train.so.8.9.7
/usr/lib64/libcudnn_cnn_infer.so.8.9.7
/usr/lib64/libcudnn_cnn_train.so.8.9.7
/usr/lib64/libcudnn_ops_infer.so.8.9.7
/usr/lib64/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              256
On-line CPU(s) list: 0-255
Thread(s) per core:  2
Core(s) per socket:  64
Socket(s):           2
NUMA node(s):        4
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7742 64-Core Processor
Stepping:            0
CPU MHz:             2250.000
CPU max MHz:         2250.0000
CPU min MHz:         1500.0000
BogoMIPS:            4500.00
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-31,128-159
NUMA node1 CPU(s):   32-63,160-191
NUMA node2 CPU(s):   64-95,192-223
NUMA node3 CPU(s):   96-127,224-255
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] optree==0.10.0
[pip3] pytorch-lightning==2.4.0
[pip3] pytorch-triton==3.0.0+dedb7bdf33
[pip3] torch==2.5.0.dev20240812+cu121
[pip3] torch_geometric==2.4.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.4.1
[pip3] torchvision==0.20.0.dev20240812+cu121
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] pytorch-lightning         2.4.0                    pypi_0    pypi
[conda] pytorch-triton            3.0.0+dedb7bdf33          pypi_0    pypi
[conda] torch                     2.5.0.dev20240812+cu121          pypi_0    pypi
[conda] torch-geometric           2.4.0                    pypi_0    pypi
[conda] torchinfo                 1.8.0                    pypi_0    pypi
[conda] torchmetrics              1.4.1                    pypi_0    pypi
[conda] torchvision               0.20.0.dev20240812+cu121          pypi_0    pypi

cc @ezyang @chauhang @penguinwu @drisspg @mikaylagawarecki @zou3519 @ydwu4 @bdhirsh @Chillee @yanboliang @BoyuanFeng

**Comment by drisspg** [comment]
I am unable to repro on H100 machine

**Comment by mishooax** [comment]
I am able to repro on H100 with torch-nightly and float16 precision

```shell
$ nvidia-smi
Thu Aug 15 14:54:12 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.86.10              Driver Version: 535.86.10    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA H100                    On  | 00000000:1B:00.0 Off |                    0 |
| N/A   38C    P0              68W / 700W |      2MiB / 65247MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+

# window.py contains the code above (torch16 working prec)
$ TORCHINDUCTOR_COMPILE_THREADS=1 python3 window.py 
Shape of output tensor: [1, 16, 40320, 16]
Segmentation fault
```

**Comment by drisspg** [comment]
@mishooax @cathalobrien would you mind trying to reproduce with the latest pytorch nightly: `pip3 install --pre torch  --index-url https://download.pytorch.org/whl/nightly/cu124`

We recently updated the version of triton shipped with pytorch and that removed alot of the segfault issues we were seeing

**Comment by cathalobrien** [comment]
Hi, this has been working for me since about 2 weeks. Apologies, should have closed this myself.  thanks for your help

**Comment by drisspg** [comment]
No worries, thanks for the response!

**Comment by zbh2047** [comment]
Hi team, it seems the problem still exists in PyTorch 2.6 official release. Compilation crashes for flex attention when turning on mixed precision. It runs normally without crash if not turning or mixed precision. Could you check if it has been merged into the official release? Thank you!

**Comment by sup3rgiu** [comment]
I'm having the same problem even without a mask.
This simple code doesn't work:

```python
import os
os.environ['TORCHINDUCTOR_COMPILE_THREADS'] = '1'

import torch
from torch.nn.attention.flex_attention import flex_attention

if __name__ == "__main__":

    torch.set_default_device("cuda")
    torch.manual_seed(0)
    torch._dynamo.config.cache_size_limit = 1000

    B = 16
    H = 8
    S = 1024
    D = 64
    data_type = torch.float16 # <-------- float32 works
    device = "cuda"

    qkv = [
        torch.randn(B, H, S, D, device=device, dtype=data_type, requires_grad=True)
        for _ in range(3)
    ]

    flex_attention_c = torch.compile(flex_attention, dynamic=False)
    out = flex_attention_c(*qkv)
```

```
python test.py
LLVM ERROR: Failed to compute parent layout for slice layout.
Aborted (core dumped)
```

Pytorch version: 2.6.0+cu124
GPU: V100 32GB

**Comment by zbh2047** [comment]
Hi team, will this bug fix be incorporated to the next version (2.6.1)? @drisspg @colesbury Thanks

**Comment by Elliot-D-Hill** [comment]
Running into the same issue in torch 2.6.0. Using `flex_attention` in a `torch.autocast` context manager for `float16` mixed precision gives the error: "RuntimeError: expected scalar type Half but found Float"
